### PR TITLE
7026262: HttpServer: improve handling of finished HTTP exchanges

### PR DESCRIPTION
**Pull request by djelinski**
This patch fixes the following issues with HttpHandler exception handling:
- The connection is no longer closed if an exception is thrown after the response is completed. As soon as the response is completed, the connection is added back to the pool, ready for reuse. Closing the connection would race with the subsequent handler.
- The stream returned by `getResponseBody` is now usable even if the response has zero-length body. Writing any data to the stream will still fail, but zero-length writes and closing the stream will now succeed.
- `ServerImpl.Exchange.reject` now sends a `Connection: close` header in addition to closing the connection

The test `B8293562` had to be adjusted to avoid `Connection: close`.
Additionally, while I was looking for a good test to copy from, I found and fixed a bug in test `B5045306`.

The new tests are passing with this patch, failing without it. Tier 1-3 clean.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-7026262](https://bugs.openjdk.org/browse/JDK-7026262): HttpServer: improve handling of finished HTTP exchanges

### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)
 * [Michael McMahon](https://openjdk.org/census#michaelm) (@Michael-Mc-Mahon - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13157/head:pull/13157` \
`$ git checkout pull/13157`

Update a local copy of the PR: \
`$ git checkout pull/13157` \
`$ git pull https://git.openjdk.org/jdk.git pull/13157/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13157`

View PR using the GUI difftool: \
`$ git pr show -t 13157`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13157.diff">https://git.openjdk.org/jdk/pull/13157.diff</a>

</details>

### CODE REVIEW

File: src/jdk.httpserver/share/classes/sun/net/httpserver/ServerImpl.java [hunk 1]
```diff
@@ -872,7 +876,7 @@ void reject (int code, String requestStr, String message) {
             rejected = true;
             logReply (code, requestStr, message);
             sendReply (
-                code, false, "<h1>"+code+Code.msg(code)+"</h1>"+message
+                code, true, "<h1>"+code+Code.msg(code)+"</h1>"+message
             );
             closeConnection(connection);
```
[the remark this answers is not in the document]
Comment (djelinski) [reply src/jdk.httpserver/share/classes/sun/net/httpserver/ServerImpl.java:881]: Good point. Will remove.

File: src/jdk.httpserver/share/classes/sun/net/httpserver/FixedLengthOutputStream.java [hunk 2]
```diff
@@ -87,12 +83,15 @@ public void close () throws IOException {
             return;
         }
         closed = true;
+        if (zerolength) {
+            // WriteFinishedEvent was sent already; nothing to do
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (djelinski) [reply src/jdk.httpserver/share/classes/sun/net/httpserver/FixedLengthOutputStream.java:89]: `sendResponseStream` could close the stream when called with length = -1; the output stream is always fixed length in that case. I'd need to modify this class to accept zero-length writes even when closed, but I think there's a precedent for that already. Will look into it.

File: test/jdk/sun/net/www/http/KeepAliveCache/B5045306.java [hunk 3]
```diff
@@ -169,7 +169,7 @@ public void handle(HttpExchange trans) {
         try {
             String path = trans.getRequestURI().getPath();
             if (path.equals("/firstCall")) {
-                port1 = trans.getLocalAddress().getPort();
+                port1 = trans.getRemoteAddress().getPort();
```
[the remark this answers is not in the document]
Comment (djelinski) [reply test/jdk/sun/net/www/http/KeepAliveCache/B5045306.java:172]: I'd rather not make any more changes to that test here.

File: src/jdk.httpserver/share/classes/sun/net/httpserver/FixedLengthOutputStream.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dfuch) [reply src/jdk.httpserver/share/classes/sun/net/httpserver/FixedLengthOutputStream.java:89]: That was just a suggestion - I was a bit bothered with the hidden interaction between FixedLengthOutputStream and sendResponseHeaders here, where FixedLengthOutputStream has to trust that sendResponseHeaders has sent the event. It looked strange. Maybe there are other possibilities - like e.g a ZeroLengthOutputStream :-)

**Review by dfuch — APPROVED** [review]
That looks much cleaner. It would be good to get @Michael-Mc-Mahon review too. But if all tests are passing and stable, it sounds good to me.

File: src/jdk.httpserver/share/classes/sun/net/httpserver/FixedLengthOutputStream.java [hunk 4]
```diff
@@ -70,10 +70,6 @@ public void write (byte[]b, int off, int len) throws IOException {
         if (closed) {
             throw new IOException ("stream closed");
         }
-        eof = (remaining == 0);
-        if (eof) {
-            throw new StreamClosedException();
-        }
```
[the remark this answers is not in the document]
Comment (djelinski) [reply src/jdk.httpserver/share/classes/sun/net/httpserver/FixedLengthOutputStream.java:76]: +1. Filed [JDK-8304873](https://bugs.openjdk.org/browse/JDK-8304873) to remove `StreamClosedException`.

File: src/jdk.httpserver/share/classes/sun/net/httpserver/FixedLengthOutputStream.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Michael-Mc-Mahon) [reply src/jdk.httpserver/share/classes/sun/net/httpserver/FixedLengthOutputStream.java]: I wonder should we move the code that sends the write finished event in this case to here? I agree this code looks pretty strange.

**Review by Michael-Mc-Mahon — APPROVED** [review]
Looks fine to me

**Pull request closed without merging**